### PR TITLE
west: runners: blackmagicprobe: use .hex/.bin file

### PR DESCRIPTION
**Pull request by MeisterBob**
If one needs the firmware to be signed (ie for mcuboot) the signed .hex or .bin file needs to be flashed to the mcu.

### CODE REVIEW

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
@MeisterBob patch looks OK but you need to fix the unit tests. See failing CI details.

**Comment by MeisterBob** [comment]
added .bin files as well and fixed the unit tests

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
@MeisterBob  I've been mostly trusting you on this PR so far, but given your last comment, I would like you to explicitly describe how you tested these changes before I approve again

**Comment by MeisterBob** [comment]
I did some more testing on this and found that the runner prefers hex over bin over elf even if the user specifies that he wants to have the bin or elf file flashed. The same problem exists in the jlink runner (#52219). Added a fix for that.

The testing procedure involved en-/disabling `CONFIG_BUILD_OUTPUT_BIN` and `CONFIG_BUILD_OUTPUT_HEX` and flash the created files with and without specifying which file I'd like to have flashed.

**Comment by MeisterBob** [comment]
Reverted that last change for an upcomming implementation of `--file` (https://github.com/zephyrproject-rtos/zephyr/issues/52262), now it's back to prefer hex over bin over elf even if `--bin-file` or `--elf-file` argument is set.

**Comment by mbolivar-nordic** [comment]
> I would like you to explicitly describe how you tested these changes before I approve again

Still waiting on this one

**Comment by MeisterBob** [comment]
the test_script:
```bash
#!/bin/bash

echo "build hex and bin"
rm build/zephyr/*.hex build/zephyr/*.bin
west build -- -DCONFIG_BUILD_OUTPUT_HEX=y -DCONFIG_BUILD_OUTPUT_BIN=y > /dev/null
echo "flashing zephyr.hex"
west flash --skip-rebuild -r blackmagicprobe --hex-file=build/zephyr/zephyr.hex
echo "flashing zephyr.bin, acually flashes zephyr.signed.hex"
west flash --skip-rebuild -r blackmagicprobe --bin-file=build/zephyr/zephyr.bin

echo "build hex only"
rm build/zephyr/*.hex build/zephyr/*.bin
west build -- -DCONFIG_BUILD_OUTPUT_HEX=y -DCONFIG_BUILD_OUTPUT_BIN=n > /dev/null
echo "flashing zephyr.hex"
west flash --skip-rebuild -r blackmagicprobe --hex-file=build/zephyr/zephyr.hex
echo "flashing zephyr.bin, acually flashes zephyr.signed.hex"
west flash --skip-rebuild -r blackmagicprobe --bin-file=build/zephyr/zephyr.bin

echo "build bin only"
rm build/zephyr/*.hex build/zephyr/*.bin
west build -- -DCONFIG_BUILD_OUTPUT_HEX=n -DCONFIG_BUILD_OUTPUT_BIN=y > /dev/null
echo "flashing zephyr.hex actually flashes zephyr.signed.bin"
west flash --skip-rebuild -r blackmagicprobe --hex-file=build/zephyr/zephyr.hex
echo "flashing zephyr.bin"
west flash --skip-rebuild -r blackmagicprobe --bin-file=build/zephyr/zephyr.bin

echo "only elf file present"
rm build/zephyr/*.hex build/zephyr/*.bin
echo "flashing zephyr.hex actually flashes zephyr.elf"
west flash --skip-rebuild -r blackmagicprobe --hex-file=build/zephyr/zephyr.hex
echo "flashing zephyr.bin actually flashes zephyr.elf"
west flash --skip-rebuild -r blackmagicprobe --bin-file=build/zephyr/zephyr.bin
```

which result in (stdout only)
```plaintext
build hex and bin
flashing zephyr.hex
-- west flash: using runner blackmagicprobe
load "build/zephyr/zephyr.hex"
flashing zephyr.bin, acually flashes zephyr.signed.hex
-- west flash: using runner blackmagicprobe
load "[...]/build/zephyr/zephyr.signed.hex"

build hex only
flashing zephyr.hex
-- west flash: using runner blackmagicprobe
load "build/zephyr/zephyr.hex"
flashing zephyr.bin, acually flashes zephyr.signed.hex
-- west flash: using runner blackmagicprobe
load "[...]/build/zephyr/zephyr.signed.hex"

build bin only
flashing zephyr.hex actually flashes zephyr.signed.bin
-- west flash: using runner blackmagicprobe
load "[...]/build/zephyr/zephyr.signed.bin" 0x1000c000
flashing zephyr.bin
-- west flash: using runner blackmagicprobe
load "build/zephyr/zephyr.bin" 0x1000c000

only elf file present
flashing zephyr.hex actually flashes zephyr.elf
-- west flash: using runner blackmagicprobe
load "[...]/build/zephyr/zephyr.elf"
flashing zephyr.bin actually flashes zephyr.elf
-- west flash: using runner blackmagicprobe
load "[...]/build/zephyr/zephyr.elf"
```

**Comment by mbolivar-nordic** [comment]
@MeisterBob thanks for the details on how you are testing this.

```
echo "flashing zephyr.bin, acually flashes zephyr.signed.hex"
west flash --skip-rebuild -r blackmagicprobe --bin-file=build/zephyr/zephyr.bin
```

Judging from this, it looks like the PR should be updated to add `--file` support to the blackmagic runner, right?

**Comment by MeisterBob** [comment]
rebased and added `--file` support.

File: scripts/west_commands/runners/blackmagicprobe.py [hunk 1]
```diff
@@ -51,16 +63,49 @@ def do_add_parser(cls, parser):
...
+            if self.file_type in (FileType.HEX, FileType.ELF):
+                flash_cmd = f'"{self.file}"'
+            elif self.file_type == FileType.BIN:
+                if self.dt_flash:
+                    flash_addr = self.flash_address_from_build_conf(self.build_conf)
+                else:
+                    flash_addr = 0
+                flash_cmd = f'"{self.file}" 0x{flash_addr:x}'
+            else:
+                err = 'Cannot flash; jlink runner only supports hex and bin files'
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/blackmagicprobe.py:81]: copy/paste error

File: scripts/west_commands/runners/blackmagicprobe.py [hunk 2]
```diff
@@ -51,16 +63,49 @@ def do_add_parser(cls, parser):
...
+                    flash_addr = self.flash_address_from_build_conf(self.build_conf)
+                else:
+                    flash_addr = 0
+                flash_cmd = f'"{self.bin_file}" 0x{flash_addr:x}'
+            elif self.elf_file is not None and os.path.isfile(self.elf_file):
+                flash_cmd = f'"{self.elf_file}"'
+            else:
+                raise ValueError('Cannot debug; no file to flash')
+
+        self.logger.info(f'Flashing file: {flash_cmd}')
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/blackmagicprobe.py:99]: This doesn't only print the file when self.file_type is BIN and self.dt_flash is true; please fix to be consistent about just printing the file name in each case.

File: scripts/west_commands/runners/blackmagicprobe.py [hunk 3]
```diff
@@ -74,37 +119,45 @@ def check_call_ignore_sigint(self, command):
...
-                            self.gdb_serial)] +
-                        self.connect_srst_disable_arg +
-                       ['-ex', "monitor swdp_scan",
-                        '-ex', "attach 1"])
+        if self.file is not None:
+            if self.file_type != FileType.ELF:
+                raise ValueError('Cannot debug; elf file required')
+            elf_name = self.file
+        elif self.elf_file is None:
+            raise ValueError('Cannot debug; elf is missing')
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/blackmagicprobe.py:127]: This used to support just attaching, with no `file`; why is this feature being removed?

File: scripts/west_commands/runners/blackmagicprobe.py [hunk 4]
```diff
@@ -74,37 +119,45 @@ def check_call_ignore_sigint(self, command):
...
 
     def bmp_debug(self, command, **kwargs):
-        if self.elf_file is None:
-            raise ValueError('Cannot debug; elf file is missing')
+        if self.file is not None:
+            if self.file_type != FileType.ELF:
+                raise ValueError('Cannot debug; elf file required')
+            elf_name = self.file
+        elif self.elf_file is None:
+            raise ValueError('Cannot debug; elf is missing')
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/blackmagicprobe.py:148]: elf file is missing; let's not make unnecessary changes here, please. We don't know what people are grepping for in scripts.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Comment by mbolivar-nordic** [comment]
> should this be closed now?

@carlescufi no; this PR is active

**Comment by carlescufi** [comment]
@mbolivar-nordic please revisit.

**Comment by mbolivar-nordic** [comment]
@MeisterBob happy new year. I'm back from vacation now. Can you share the updated test procedure you used?
